sway-lib-core/src/ops.sw

```diff
@@ -173,17 +173,38 @@ impl Subtract for u64 {
     }
 }
 
-// unlike addition, underflowing subtraction does not need special treatment
-// because VM handles underflow
 impl Subtract for u32 {
     fn subtract(self, other: Self) -> Self {
-        __sub(self, other)
+        // any non-64-bit value is compiled to a u64 value under-the-hood
+        // constants (like Self::max() below) are also automatically promoted to u64
+        let res = __sub(self, other);
```

Comment:
I think doing `__sub` like this is an "undefined behaviour" because it can generate an "invalid u32". We can decide to fix this in the compiler somehow. 

Given that you are already doing a comparison, why not just compare `self` to `other` directly?

```sway
if __gt(other, self) {
     ...
} else {
    // SAFE: self is greater than other
    __sub(self, other)
}
```